google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java

```diff
@@ -0,0 +1,3482 @@
+          ErrorCode.INVALID_ARGUMENT, "Timestamp parse error", e);
+    }
+  }
+
+  /** Convert a cloud Date to a Date proto. */
+  private static int daysFromDate(Date date) {
+    return (int) LocalDate.of(date.getYear(), date.getMonth(), date.getDayOfMonth()).toEpochDay();
+  }
+
+  /** Convert a Date proto to a cloud Date. */
+  private static Date dateFromDays(int daysSinceEpoch) {
+    LocalDate localDate = LocalDate.ofEpochDay(daysSinceEpoch);
+    return Date.fromYearMonthDay(
+        localDate.getYear(), localDate.getMonthValue(), localDate.getDayOfMonth());
+  }
+
+  @Nullable
+  private static ByteString toByteString(@Nullable ByteArray byteArray) {
+    if (byteArray == null) {
+      return null;
+    }
+    return ByteString.copyFrom(byteArray.toByteArray());
+  }
+
+  @Nullable
+  private static ByteArray toByteArray(@Nullable ByteString byteString) {
+    if (byteString == null) {
+      return null;
+    }
+    return ByteArray.copyFrom(byteString.toByteArray());
+  }
+
+  /** Convert a list of nullable value to another type. */
+  private static <S, T> List<T> unmarshallValueList(
+      List<Boolean> isNullList, List<S> valueList, Function<S, T> converter) {
+    List<T> newValueList = new ArrayList<>(valueList.size());
+    if (isNullList.isEmpty()) {
+      for (S value : valueList) {
+        newValueList.add(converter.apply(value));
+      }
+    } else {
+      for (int i = 0; i < valueList.size(); ++i) {
+        newValueList.add(isNullList.get(i) ? null : converter.apply(valueList.get(i)));
+      }
+    }
+    return newValueList;
+  }
+
+  /** Insert null into valueList according to isNullList. */
+  private static <S> List<S> unmarshallValueList(List<Boolean> isNullList, List<S> valueList) {
+    return unmarshallValueList(isNullList, valueList, element -> element);
+  }
+
+  /** Convert a Struct proto to a cloud Struct. */
+  private static com.google.cloud.spanner.Struct structProtoToCloudStruct(
+      com.google.spanner.v1.Type type, com.google.spanner.executor.v1.ValueList structValue) {
+    List<com.google.spanner.executor.v1.Value> fieldValues = structValue.getValueList();
+    List<StructType.Field> fieldTypes = type.getStructType().getFieldsList();
+
+    if (fieldTypes.size() != fieldValues.size()) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INVALID_ARGUMENT,
+          "Mismatch between number of expected fields and specified values for struct type");
+    }
+
+    com.google.cloud.spanner.Struct.Builder builder = com.google.cloud.spanner.Struct.newBuilder();
+    for (int i = 0; i < fieldTypes.size(); ++i) {
+      builder
+          .set(fieldTypes.get(i).getName())
+          .to(valueProtoToCloudValue(fieldTypes.get(i).getType(), fieldValues.get(i)));
+    }
+    return builder.build();
+  }
+
+  /** Convert a Type proto to a cloud Type. */
+  private static com.google.cloud.spanner.Type typeProtoToCloudType(
```

Comment:
Can we re-use https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/java/com/google/cloud/spanner/Type.java#L442?

Reply:
This API is not public.